src/python.cpp

```diff
@@ -942,6 +942,85 @@ SEXP py_to_r(PyObject* x, bool convert) {
   }
 }
 
+// [[Rcpp::export]]
+SEXP py_get_formals(PyObjectRef func, bool convert) {
+  PyObjectPtr inspect(py_import("inspect"));
+  if (inspect.is_null()) stop(py_fetch_error());
+  PyObjectPtr get_signature(PyObject_GetAttrString(inspect.get(), "signature"));
+  if (get_signature.is_null()) stop(py_fetch_error());
+  PyObjectPtr signature(PyObject_CallFunctionObjArgs(get_signature.get(), func.get(), NULL));
+  // if we could not get a signature, return an empty one
+  if (signature.is_null() || PyErr_Occurred()) {
+    // TODO: restrict to ValueError
+    if (PyErr_Occurred()) PyErr_Clear();
+    return R_NilValue;
+  }
+  PyObjectPtr param_dict(PyObject_GetAttrString(signature.get(), "parameters"));
+  if (param_dict.is_null()) stop(py_fetch_error());
+
+  PyObjectPtr param_values(PyObject_GetAttrString(param_dict.get(), "values"));
+  if (param_values.is_null()) stop(py_fetch_error());
+  PyObjectPtr params(PyObject_CallFunctionObjArgs(param_values.get(), NULL, NULL));
+  if (params.is_null()) stop(py_fetch_error());
+  PyObjectPtr param_iter(PyObject_GetIter(params.get()));
+  if (param_iter.is_null()) stop(py_fetch_error());
+
+  // Static properties of the Parameter class
+  PyObjectPtr param_class(PyObject_GetAttrString(inspect.get(), "Parameter"));
+  if (param_class.is_null()) stop(py_fetch_error());
+  PyObjectPtr empty_param(PyObject_GetAttrString(param_class.get(), "empty"));
+  if (empty_param.is_null()) stop(py_fetch_error());
+  PyObjectPtr var_pos(PyObject_GetAttrString(param_class.get(), "VAR_POSITIONAL"));
+  if (var_pos.is_null()) stop(py_fetch_error());
+  PyObjectPtr var_kw(PyObject_GetAttrString(param_class.get(), "VAR_KEYWORD"));
+  if (var_kw.is_null()) stop(py_fetch_error());
+  PyObjectPtr kw_only(PyObject_GetAttrString(param_class.get(), "KEYWORD_ONLY"));
+  if (kw_only.is_null()) stop(py_fetch_error());
+
+  Rcpp::Pairlist formals;
+  PyObject* param;
+  bool var_encountered = false;
+  while ((param = PyIter_Next(param_iter.get()))) {
```

Comment:
`PyIter_Next()` returns a new reference. Does that reference need to be cleaned up here?

Reply:
Thank you, right! Done in 54389e84a547a5207c0098d41362cb64750f047e

IDK if there’s a more idiomatic way to do this in your codebase: Maybe I should use `py_iter_next`?

Reply:
I’m wrapping the thing in a PyObjectPtr now!